install-server.sh

```diff
@@ -63,6 +63,8 @@ cmdline() {
     if [[ -z "$PRIVATE_KEY" ]]; then
         read -rp "Please provide the path of the runner private key: " PRIVATE_KEY </dev/tty
     fi
+    # Replace tilde with the current home:
+    PRIVATE_KEY="${PRIVATE_KEY/#\~/$HOME}"
```

Comment:
@tomschr  thanks! I would use something like `PRIVATE_KEY="$(cd "$(dirname "$PRIVATE_KEY")"; pwd -P)/$(basename "$PRIVATE_KEY")"`  to make it even more generic and able to expand tilde and relative paths as well.

Reply:
@fabriziosestito thanks for the reply!

To be honest, I don't really like the `cd` and `basename`, :wink: sorry. Two commands executed just to replace the tilde? Could we replace it with at least a single command?

After searching a bit, I found `realpath`:

```
$ realpath ~/.ssh/id_rsa
/home/tux/.ssh/id_rsa
```

You could even add the options `--logical` to resolve any `..` components or to resolve symlinks if you like. That would work with relative paths. Find all details in the [realpath man page](https://man7.org/linux/man-pages/man1/realpath.1.html).

With that, I would suggest:

```
PRIVATE_KEY=$(realpath "$PRIVATE_KEY")
```

Would that be more in line with your needs?

Reply:
@tomschr `realpath` is sometimes not installed in bare-minimum systems

Reply:
@fabriziosestito /usr/bin/realpath is part of GNU coreutils. Do you mean that it's possible to have minimal systems without coreutils ?

Reply:
@sionescu @tomschr  thanks! if it is in coreutils let's go for it, i had very outdated memories that it was packaged separately

Reply:
More precisely, `realpath` is part of coreutils since version 8.15, [released on 2012-01-06](https://savannah.gnu.org/forum/forum.php?forum_id=7076), so it's safe to depend on it.

Reply:
Ok, great! :+1: 

@fabriziosestito @sionescu There is only one question left: how do we deal with invalid user input?

If I'm not mistaken, the script does not have any checks yet, right? I think we should add that too to avoid non-existing paths.

According to the man page, the option `-e`/`--canonicalize-existing` checks that "all components of the path must exist". If not, the result will be empty and you get a return code of -1. We could either check for the return code or against the empty string. Any preferences?

Here is an example with checking against an empty string:

```bash
TMP=$(realpath -q -e "$PRIVATE_KEY")
if [ -z "$TMP" ]; then
   echo "Path '$PRIVATE_KEY' to private SSH key does not exist, please try again."
   exit 1
fi
PRIVATE_KEY="$TMP"
```

Reply:
We can check the command's return value, and since the interpreter is /bin/bash we can use the better-behaved operator "[[". Also, to improve code clarity, I suggest always delimiting variable references with curly braces, except for the one-letter built-ins.

```shell
TMP=$(realpath -q -e "${PRIVATE_KEY}")
if [[ $? > 0 ]]; then
   echo "Path '${PRIVATE_KEY}' to private SSH key does not exist, please try again."
   exit 1
fi
```

Reply:
@tomschr this is perfect thanks! Just couple of nitpicks: I would just rename the `tmp` variable in something more meaningful `private_key_absolute_path` so we document what we are doing as well, and maybe change the message "to private SSH key does not exist, please provide a valid path."

Reply:
Thank you! I've combined the two suggestions from @sionescu and @fabriziosestito and pushed it.
Check it out and let me know if there are more things to change. Thank you! :+1: